### ISSUE TITLE
Pass page context variables into static queries

### ISSUE THREAD

**Issue by daaru00** [gridsome/gridsome#1107]
## Summary

If I understand well, at this moment, static query did not accept any input variable at all. In some cases in not possible to get data with a fixed query, for example when website is localized. Need to access page context variables to retrieve the correct data in the current context.

### Basic example

Creating two different page that use the same component but with a different locale:
```js
//gridsome.server.js

module.exports = function (api) {
  api.createPages(({ createPage }) => {
     // Create an english page
     createPage({
        path: `/en/page/`,
        component: './src/templates/Page.vue',
        context: {
          locale: 'en-gb'
        }
      })

     // Create an italian page
     createPage({
        path: `/it/page/`,
        component: './src/templates/Page.vue',
        context: {
          locale: 'it-it'
        }
      })
  })
}
```

Inside `Page.vue` is included a component `./src/components/Sidebar.vue` that has the static query:
```html
<static-query>
query($locale: String) {
   SidebarItem(id: "sidebar", language: $locale) {
      content {
        _uid
        name
        subtitle
      }
   }
}
</static-query>
```
the `id` parameter can be hard-coded but the locale not. Is too awkward to clone each component for each language (`./src/components/SidebarEN.vue`, `./src/components/SidebarIT.vue`, ..) and include it dynamically (or using v-if based on locale).

### Motivation

Some datas cannot be retrieved with only hard-coded parameters but need to depends from page context.

**Comment by thetre97** [comment]
@daaru00 Could you not query that data in `Page.vue` with `<page-query>`, and pass as a prop to the sidebar component?

**Comment by daaru00** [comment]
@thetre97 nope, I can't merge it together. Sidebar has too many fields so I need to retrieve it separately.

**Comment by SuiOni** [comment]
Bump

**Comment by thetre97** [comment]
I'm not sure that this will happen, as the point of it being called a static query is because it doesn't receive variables.

A couple of workarounds is to either query the data in a `page-query` and pass down via props, or fetch all variations and use a computed property to filter and get the right one for the current locale.

**Comment by SuiOni** [comment]
The first one is the natural way for me, just makes the query quite long. The second option would increase the site size as the data is preshiped and the lubricated right?

________________________________
Von: Travis Reynolds <notifications@github.com>
Gesendet: Monday, February 8, 2021 8:37:26 PM
An: gridsome/gridsome <gridsome@noreply.github.com>
Cc: Sak <s.zirai@live.de>; Comment <comment@noreply.github.com>
Betreff: Re: [gridsome/gridsome] Pass page context variables into static queries (#1107)


I'm not sure that this will happen, as the point of it being called a static query is because it doesn't receive variables.

A couple of workarounds is to either query the data in a page-query and pass down via props, or fetch all variations and use a computed property to filter and get the right one for the current locale.

**Comment by thetre97** [comment]
Correct, it would - although it shouldn't be too much. This query for example
```graphql
query($locale: String) {
   SidebarItem(id: "sidebar", language: $locale) {
      content {
        _uid
        name
        subtitle
      }
   }
}
```
Wouldn't generate a huge amount of JSON.

Hopefully fragments & importing graphql files will come soon...

**Comment by jthomerson** [comment]
Another use case for this: I'm using `@gridsome/vue-remark`. I want to create a "related posts" component that editors can embed in their Markdown somewhere to advertise related posts somewhere within the content. They should be able to pass the component some value, for example `<related-posts :tag="tags[0]" />` to find other posts with the same tag. But then the component needs to do the query because the person writing the page (the very generic "MarkdownPost" page template) has no idea what components the Markdown author will write.

Any ideas on how to accomplish something like this?

**Comment by rasco** [comment]
also interested in this. it seems most static site generators can't do this, but it would be a killer feature, because it lets you move the data sourcing complexity into the components responsible for that data.
